appnav/src/main/kotlin/io/element/android/appnav/di/DefaultSyncOrchestrator.kt

```diff
@@ -0,0 +1,148 @@
+import io.element.android.services.appnavstate.api.SyncOrchestrator
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import timber.log.Timber
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+class DefaultSyncOrchestrator @AssistedInject constructor(
+    @Assisted matrixClient: MatrixClient,
+    private val baseCoroutineScope: CoroutineScope = matrixClient.sessionCoroutineScope,
+    private val appForegroundStateService: AppForegroundStateService,
+    private val networkMonitor: NetworkMonitor,
+    private val dispatchers: CoroutineDispatchers,
+) : SyncOrchestrator {
+    @AssistedFactory
+    interface Factory {
+        fun create(matrixClient: MatrixClient): DefaultSyncOrchestrator
+    }
+
+    private val syncService = matrixClient.syncService()
+
+    private val initialSyncMutex = Mutex()
+
+    private var coroutineScope: CoroutineScope? = null
+
+    private val tag = "SyncOrchestrator"
+
+    private val started = AtomicBoolean(false)
+
+    /**
+     * Starting observing the app state and network state to start/stop the sync service.
+     *
+     * Before observing the state, a first attempt at starting the sync service will happen if it's not already running.
+     */
+    @OptIn(FlowPreview::class)
+    override fun start() {
+        if (!started.compareAndSet(false, true)) {
+            Timber.tag(tag).d("already started, exiting early")
+            return
+        }
+
+        Timber.tag(tag).d("start observing the app and network state")
+
+        if (syncService.syncState.value != SyncState.Running) {
+            Timber.tag(tag).d("initial startSync")
+            baseCoroutineScope.launch(dispatchers.io) {
+                try {
+                    initialSyncMutex.lock()
+                    syncService.startSync()
+
+                    // Wait until it's running
+                    syncService.syncState.first { it == SyncState.Running }
+                } finally {
+                    initialSyncMutex.unlock()
+                }
+            }
+        }
+
+        coroutineScope = CoroutineScope(baseCoroutineScope.coroutineContext + CoroutineName(tag) + dispatchers.io)
+
+        coroutineScope?.launch {
+            // Wait until the initial sync is done, either successfully or failing
+            initialSyncMutex.lock()
+
+            combine(
+                // small debounce to avoid spamming startSync when the state is changing quickly in case of error.
```

Comment:
Maybe the debounced can be moved after the combine block (before .distinctUntilChanged())?

Reply:
This causes issues in the tests, since it's preventing some intermediate states from being generated. I'd rather keep it were it is, if possible.

Reply:
Can you try with this patch?

[moveDebounce.patch](https://github.com/user-attachments/files/18691006/moveDebounce.patch)